Mathlib/RingTheory/Flat/Basic.lean

```diff
@@ -35,44 +37,52 @@ This result is not yet formalised.
 * `Module.Flat.directSum`: arbitrary direct sums of flat modules are flat
 * `Module.Flat.of_free`: free modules are flat
 * `Module.Flat.of_projective`: projective modules are flat
+* `Module.Flat.iff_rTensor_preserves_injective_linearMap`: a module is flat iff tensoring preserves
+  injectiveness.
+
+## Implementation notes
+In `Module.Flat.iff_rTensor_preserves_injective_linearMap`, we require that the universe level of
+the ring is lower than or equal to that of the module. This requirement is to make sure ideals of
+the ring can be lifited to the universe of the module. It is unclear if this lemma also holds
+when module lives in a lower universe.
+
+This requirement also appears in `Algebra/ModuleCat/Injective`.
+There are some ideas proposed by Junyan Xu about potentially circumvent at
+[here](https://github.com/leanprover-community/mathlib4/pull/8905#discussion_r1428509361)
```

Comment:
Five comments:

1. The title of #8905 is misleading and I regret not catching it before it's merged: it should say that if the module is injective in the ring's universe (or higher), then it's injective in all universes (lower or higher). What's true in general is rather the converse: if a module is injective in a universe then it's also injective in all lower universes. The reason we can pass from the ring's universe to an arbitrary universe is Baer's criterion, which says it suffices to take the inclusions of ideals in the ring as the test morphisms, even though the definition of Module.Injective requires all injective linear maps between modules in a universe as test morphisms.

2. Our definition of Module.Flat is exactly analogous to Module.Baer: both use the inclusions of ideals in the ring as test morphisms. In fact, we can show that M is flat iff M* is Baer, and that if `M : Type v`, then `M ⊗[R] -` preserves injective linear maps in between modules in universe v iff M* satisfies `Module.Injective R` (and probably is also equivalent to that Tor^n(M,-) vanishes in `ModuleCat.{v} R` for n > 0).

3. I'm now convinced that there are modules that are injective in their universe but not in the ring's universe (i.e. doesn't satisfy Module.Baer), which means that my "circumvention" can't be made to work). We just need to give an example of a module M whose injective hull can't fit in M's universe; here by an injective hull of M I mean an essential extension of M that is injective in all universes, which exists by the [usual construction](https://stacks.math.columbia.edu/tag/08Y3), and it is unique (up to non-unique isomorphism). Since M itself is an essential extension of M, it would be isomorphic to its injective hull if it were injective, but this can't happen if the injective hull doesn't fit in the universe of M.
I believe the following example works: take a field `k` in `Type 0` (e.g. Rat) and take a type σ that is too large to fit in Type 0, and consider `R := MvPolynomial σ k`. Consider the maximal ideal `m` generated by all variables, and let `M := R/m ≃ k` (i.e., R acts on k such that every variable acts as 0). I believe the injective hull E of the R-module M has cardinality at least #σ. In fact, if σ were finite, say σ = Fin n, then [we know](https://math.stackexchange.com/questions/4579568/injective-hull-of-k-as-kx-1-x-n-module) that $E\cong k[x_1^{-1}, \dots, x_n^{-1}]$. In general, $1 \in k \cong M \subseteq E$ must be divisible by all variables, and we can probably show $1/x_i \ne 1/x_j$ in E if $i \ne j \in \sigma$, but I haven't worked out the argument. We could also use [Matlis duality](https://stacks.math.columbia.edu/tag/08Z6) if σ were finite (warning: "Before about 1960 many authors required that a local ring be (left and right) Noetherian" according to Wikipedia), which implies that Hom_R(E,E) is isomorphic to the completion of R at m. When σ is infinite, we can probably still show that R injects into Hom_R(E,E), so E can't fit in Type 0 because if it does, then E → E, Hom_R(E,E), R, and therefore σ also fit in Type 0.

4. It's unclear whether we could give an example of an R-module M such that `M ⊗[R] -` preserves injective linear maps in M's universe but not in R's universe (i.e. M doesn't satisfy Module.Flat). For this we need to produce an R-module N such that the injective hull of the character module N* doesn't fit in M's universe. The example in 3. probably isn't the character module of some module.

5. there's a typo: "lifited".

Reply:
re 2: I think you are right, I did it in Lean3, I think this for $n = 1$ by applying the snake lemma.
re 3: I am convinced by your argument, it would be a really cool project in the `CounterExample` folder. I think point 3 should go into `ModuleCat/Injective.lean` as an implementation note at least.
re 5: fixed

Reply:
re 1: Indeed, the definition of Module.Injective uses test objects in the same universe as M, so "being injective in a lower universe" doesn't literally make sense unless you Shrink M to a lower universe using a Small assumption. What I had in mind is something like 
```
∀ ⦃X Y : Type v'⦄ [inst : AddCommGroup X] [inst_1 : AddCommGroup Y] [inst_2 : Module R X] [inst_3 : Module R Y]
  (f : X →ₗ[R] Y), Function.Injective ⇑f → ∀ (g : X →ₗ[R] Q), ∃ (h : Y →ₗ[R] Q), ∀ (x : X), h (f x) = g x
```
where v' is smaller than M's universe.